crates/driver/src/domain/competition/bad_tokens/mod.rs

```diff
@@ -0,0 +1,123 @@
+    Unsupported,
+}
+
+#[derive(Default)]
+pub struct Detector {
+    /// manually configured list of supported and unsupported tokens. Only
+    /// tokens that get detected incorrectly by the automatic detectors get
+    /// listed here and therefore have a higher precedence.
+    hardcoded: HashMap<eth::TokenAddress, Quality>,
+    simulation_detector: Option<simulation::Detector>,
+    metrics: Option<metrics::Detector>,
+}
+
+impl Detector {
+    /// Hardcodes tokens as (un)supported based on the provided config. This has
+    /// the highest priority when looking up a token's quality.
+    pub fn new(config: HashMap<eth::TokenAddress, Quality>) -> Self {
+        Self {
+            hardcoded: config,
+            ..Default::default()
+        }
+    }
+
+    /// Enables detection of unsupported tokens via simulation based detection
+    /// methods.
+    pub fn with_simulation_detector(&mut self, detector: simulation::Detector) -> &mut Self {
+        self.simulation_detector = Some(detector);
+        self
+    }
+
+    /// Enables detection of unsupported tokens based on heuristics.
+    pub fn with_heuristic_detector(&mut self) -> &mut Self {
+        self.metrics = Some(metrics::Detector);
+        self
+    }
+
+    /// Removes all unsupported orders from the auction.
+    pub async fn filter_unsupported_orders_in_auction(&self, mut auction: Auction) -> Auction {
+        let now = Instant::now();
+
+        let filtered_orders = futures::stream::iter(auction.orders.into_iter())
+            .filter_map(move |order| async move {
+                let sell = self.get_token_quality(order.sell.token, now);
+                let buy = self.get_token_quality(order.sell.token, now);
+                match (sell, buy) {
+                    // both tokens supported => keep order
+                    (Some(Quality::Supported), Some(Quality::Supported)) => Some(order),
+                    // at least 1 token unsupported => drop order
+                    (Some(Quality::Unsupported), _) | (_, Some(Quality::Unsupported)) => None,
+                    // sell token quality is unknown => keep order if token is supported
+                    (None, _) => {
+                        let Some(detector) = &self.simulation_detector else {
+                            // we can't determine quality => assume order is good
+                            return Some(order);
+                        };
+                        let quality = detector.determine_sell_token_quality(&order, now).await;
+                        (quality == Some(Quality::Supported)).then_some(order)
+                    }
+                    // buy token quality is unknown => keep order (because we can't
+                    // determine quality and assume it's good)
+                    (_, None) => Some(order),
+                }
+            })
+            .collect::<Vec<_>>()
+            .await;
+        auction.orders = filtered_orders;
```

Comment:
Should we log the filtered-out orders here? Seems like it would be painful to understand why an order wasn't executed.

Reply:
I initially left it out because I'm worried that those logs will be pretty noisy but to make oncall easier we should probably have them anyway.

Reply:
Maybe we can avoid logging `0x0000...` order, which is used in the fake auction? That should probably reduce the amount of logs.

Reply:
The fake auctions used for quoting don't pass through this logic. Maybe they should (not entirely clear IMO) but I would do that as a follow up - if at all.